### ISSUE TITLE
Error on subquery with aggregate

### ISSUE THREAD

**Issue by LorenzBuehmann** [ontop/ontop#405]
Hi Ontop devs!

I'm getting the following exception (latest develop version used): 

```
it.unibz.inf.ontop.iq.exception.InvalidIntermediateQueryException: The following non-projected variables [x] are appearing in different children of JOIN OR2(STRICT_EQ2(max0,max1),OR2(IS_NULL(max0),IS_NULL(max1))):   [ AGGREGATE [] [max0/SP_MAX(y)]    INTENSIONAL triple(x,,y) ,  INTENSIONAL triple(x,,max1)]
it.unibz.inf.ontop.exception.OntopReformulationException
it.unibz.inf.ontop.answering.reformulation.impl.QuestQueryProcessor.reformulateIntoNativeQuery(QuestQueryProcessor.java:144)
```

Query
```
prefix ex:	<http://www.example.org/schema#>
prefix in:	<http://www.example.org/instance#>

select ?x ?max where {
{select (max(?y) as ?max) where {?x ex:p ?y} } 
?x ex:p ?max
}
```

SQL

```
CREATE TABLE IF NOT EXISTS "t1"
("s" VARCHAR NOT NULL,"o" BIGINT NOT NULL, PRIMARY KEY ("s","o"))
```

Mappings
```
[ <http://www.w3.org/ns/r2rml#logicalTable>
          [ <http://www.w3.org/ns/r2rml#tableName>
                    "\"t1\"" ] ;
  <http://www.w3.org/ns/r2rml#predicateObjectMap>
          [ <http://www.w3.org/ns/r2rml#objectMap>
                    [ <http://www.w3.org/ns/r2rml#column>
                              "\"o\"" ;
                      <http://www.w3.org/ns/r2rml#datatype>
                              <http://www.w3.org/2001/XMLSchema#integer>
                    ] ;
            <http://www.w3.org/ns/r2rml#predicate>
                    <http://www.example.org/schema#p>
          ] ;
  <http://www.w3.org/ns/r2rml#subjectMap>
          [ <http://www.w3.org/ns/r2rml#column>
                    "\"s\"" ;
            <http://www.w3.org/ns/r2rml#termType>
                    <http://www.w3.org/ns/r2rml#IRI>
          ]
] .
```

Looks like an issue with aggregate in a subquery possibly? (note, this is a test case from the W3C SPARQL 1.1 test suite "sq08 - Subquery with aggregate")

So can you reproduce this issue? Or is it something on my end?

Cheers

**Comment by bcogrel** [comment]
Hi Lorenz,

I think the problem comes from the fact the variable ?x appears both within the sub-query (but is not projected) and outside. 
We could made Ontop more robust to that but this will require a bit of work.

This can be avoided by reformulating the query as follows:

```sparql
prefix ex:	<http://www.example.org/schema#>
prefix in:	<http://www.example.org/instance#>

select ?x ?max where {
{select (max(?y) as ?max) where {?s ex:p ?y} } 
?x ex:p ?max
}
```

Best,
Benjamin

**Comment by bcogrel** [comment]
Actually, I didn't manage to reproduce your issue. Which branch did you use? `develop` or `version4`?

`develop` corresponds to 1.18.2 and is therefore quite old. We should probably remove it.
`version4` is the branch we use the development of the next major release (4.1.0).

**Comment by LorenzBuehmann** [comment]
Hi Benjamin,

well I'm dumb ... clearly I meant latest `version4` branch. I'm always thinking that `develop` is the latest development branch ...

Just to recap, this query is a test case from the SPARQL 1.1 conformance test suite. I'm wondering why this test doesn't fail in Ontop itself. Let me check and compare the debug output in the reformulation part. I'll come back to you once I got more insights. I forked Ontop but didn't touch the core (i thought), but who knows ...

Cheers

**Comment by bcogrel** [comment]
Hi Lorenz,

Sorry I replied too quickly yesterday without properly investigating. I have found where the problem is coming from.

So, the test "sq08" is running fine until you set the property `ontop.testMode` to true. 

With this option on, Ontop performs extra checks and reports a violation of the assumptions holding in our internal algebra. Here this violation is without consequence, but it could avoided by renaming the variable below the aggregation when the SPARQL query is translated into our internal algebra.

TODO:
- [ ] Enable the test mode in SPARQL compliance tests
- [ ] Rename conflicting non-projected variables of an aggregation sub-tree. 

Thanks for reporting the issue!
Best,
Benjamin
